services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.security.JaasUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kmf.common.Utils.ZK_CONNECTION_TIMEOUT_MS;
+import static com.linkedin.kmf.common.Utils.ZK_SESSION_TIMEOUT_MS;
+
+import scala.collection.Seq;
+
+
+/**
+ * Runs this periodically to rebalance the monitored topic across brokers and to reassign _electedLeaders to brokers so that the
+ * monitored topic is sampling all the brokers evenly.
+ */
+public class TopicRebalancer implements Service  {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TopicRebalancer.class);
+
+  /**
+   * The state of the monitored topic.
+   */
+  static class TopicState {
+
+    Map<Integer, Integer> _brokerToPartitionCount = new HashMap<>();
+    //Assigned _electedLeaders tracks if there is some partition for which it is the preferred leader which could be none.
+    Set<Integer> _preferredLeaders = new HashSet<>();
+    Set<Integer> _electedLeaders = new HashSet<>();
+    Collection<Broker> _allBrokers;
+    List<PartitionInfo> _partitionInfo;
+
+    TopicState(Map<Integer, Integer> brokerToPartitionCount, Set<Integer> electedLeaders,
+      Set<Integer> preferredLeaders, List<PartitionInfo> partitionInfo, Collection<Broker> allBrokers) {
+      this._brokerToPartitionCount = brokerToPartitionCount;
+      this._preferredLeaders = preferredLeaders;
+      this._electedLeaders = electedLeaders;
+      this._allBrokers = allBrokers;
+      this._partitionInfo = partitionInfo;
+    }
+
+    /**
+     *
+     * @param expectedRatio the expected partition to broker ratio
+     * @return true if the current partition to broker ratio is less than the expected ratio
+     */
+    boolean partitionsLow(double expectedRatio) {
+      double actualRatio = ((double) _partitionInfo.size()) / _allBrokers.size();
+      return actualRatio < expectedRatio;
+    }
+
+    /**
+     * @return true if broker is not the preferred leader for at least one partition.
+     */
+    boolean brokerMissingPartition() {
+
+      for (Broker broker : _allBrokers) {
+        if (!_brokerToPartitionCount.containsKey(broker.id())) {
+          return true;
+        }
+
+        if (!_preferredLeaders.contains(broker.id())) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    /**
+     * @return true if at least one broker is not elected leader for at least one partition.
+     */
+    boolean brokerNotElectedLeader() {
+      for (Broker broker : _allBrokers) {
+        if (!_electedLeaders.contains(broker.id())) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+  }
+
+  private class TopicRebalancerRunnable implements Runnable {
+    @Override
+    public void run() {
+      try {
+        TopicState topicState = topicState();
+        if (topicState.brokerNotElectedLeader() || topicState.brokerMissingPartition() ||
```

Comment:
I like to separate out the detection of a problem from the action that is taken.